homeassistant/components/snooz/fan.py

```diff
@@ -0,0 +1,155 @@
+
+from homeassistant.components.fan import ATTR_PERCENTAGE, FanEntity, FanEntityFeature
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_ADDRESS, STATE_OFF, STATE_ON
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.restore_state import RestoreEntity
+
+from .const import DOMAIN
+from .models import SnoozConfigurationData
+
+# transitions logging is pretty verbose, so only enable warnings/errors
+logging.getLogger("transitions.core").setLevel(logging.WARNING)
+
+_LOGGER = logging.getLogger(__name__)
+_LOGGER.setLevel(logging.DEBUG)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up Snooz device from a config entry."""
+
+    address: str = entry.data[CONF_ADDRESS]
+    data: SnoozConfigurationData = hass.data[DOMAIN][entry.entry_id]
+
+    async_add_entities(
+        [
+            SnoozFan(
+                hass,
+                data.device.display_name,
+                address,
+                data.device,
+            )
+        ]
+    )
+
+
+class SnoozFan(FanEntity, RestoreEntity):
+    """Fan representation of a Snooz device."""
+
+    def __init__(self, hass, name: str, address: str, device: SnoozDevice) -> None:
+        """Initialize a Snooz fan entity."""
+        self.hass = hass
+        self._address = address
+        self._device = device
+        self._attr_unique_id = address
+        self._attr_supported_features = FanEntityFeature.SET_SPEED
+        self._attr_name = name
+        self._attr_is_on = None
+        self._attr_should_poll = False
+        self._attr_percentage = None
+
+    def _write_state_changed(self) -> None:
+        # cache state for restore entity
+        if not self.assumed_state:
+            self._attr_is_on = self._device.state.on
+            self._attr_percentage = self._device.state.volume
+
+        self.async_write_ha_state()
+
+    def _on_connection_status_changed(self, new_status: SnoozConnectionStatus) -> None:
+        self._write_state_changed()
+
+    def _on_device_state_changed(self, new_state: SnoozDeviceState) -> None:
+        self._write_state_changed()
+
+    async def async_added_to_hass(self) -> None:
+        """Restore state and subscribe to device events."""
+        await super().async_added_to_hass()
+
+        if last_state := await self.async_get_last_state():
+            self._attr_is_on = (
+                last_state.state == STATE_ON
+                if last_state.state in (STATE_ON, STATE_OFF)
+                else None
+            )
+            self._attr_percentage = last_state.attributes.get(ATTR_PERCENTAGE)
+
+        self.async_on_remove(self._subscribe_to_device_events())
+
+    def _subscribe_to_device_events(self) -> Callable[[], None]:
```

Comment:
```suggestion
    @callback
    def _async_subscribe_to_device_events(self) -> Callable[[], None]:
```

Since Home Assistant has both sync and async function in the codebase we usually prefix any callback functions that are safe to call in async with `async`

The `@callback` decorator is also a hint to Home Assistant to not run the function in the executor